tests/src/test/scala-2.12/cats/tests/ScalaVersionSpecific.scala

```diff
@@ -1,6 +1,57 @@
 package cats.tests
 
+import cats.laws.discipline.{ExhaustiveCheck, MiniInt}
+import cats.laws.discipline.MiniInt._
+import cats.laws.discipline.eq._
+import cats.kernel.{Eq, Order}
+import cats.syntax.either._
+
 trait ScalaVersionSpecificFoldableSuite
 trait ScalaVersionSpecificParallelSuite
 trait ScalaVersionSpecificRegressionSuite
 trait ScalaVersionSpecificTraverseSuite
+
+trait ScalaVersionSpecificAlgebraInvariantSuite {
+  protected val integralForMiniInt: Integral[MiniInt] = new Integral[MiniInt] {
+    def compare(x: MiniInt, y: MiniInt): Int = Order[MiniInt].compare(x, y)
+    def plus(x: MiniInt, y: MiniInt): MiniInt = x + y
+    def minus(x: MiniInt, y: MiniInt): MiniInt = x + (-y)
+    def times(x: MiniInt, y: MiniInt): MiniInt = x * y
+    def negate(x: MiniInt): MiniInt = -x
+    def fromInt(x: Int): MiniInt = MiniInt.unsafeFromInt(x)
+    def toInt(x: MiniInt): Int = x.toInt
+    def toLong(x: MiniInt): Long = x.toInt.toLong
+    def toFloat(x: MiniInt): Float = x.toInt.toFloat
+    def toDouble(x: MiniInt): Double = x.toInt.toDouble
+    def quot(x: MiniInt, y: MiniInt): MiniInt = MiniInt.unsafeFromInt(x.toInt / y.toInt)
+    def rem(x: MiniInt, y: MiniInt): MiniInt = MiniInt.unsafeFromInt(x.toInt % y.toInt)
+  }
+
+  implicit protected def eqNumeric[A: Eq: ExhaustiveCheck]: Eq[Numeric[A]] = {
+
+    // These allow us to capture the cases where operations on Numeric throw (eg when causing an overflow). These are
+    // represented by `None` and allow us to compare two Numeric instances, verifying that when one throws, the other
+    // also throws.
+    def makeUnaryFnSafe[X, R](f: X => R): X => Option[R] =
+      x => Either.catchOnly[IllegalArgumentException](f(x)).toOption
+    def makeBinaryFnSafe[X, Y, R](f: (X, Y) => R): (X, Y) => Option[R] =
+      (x, y) => Either.catchOnly[IllegalArgumentException](f(x, y)).toOption
```

Comment:
Note that here and in a couple of other places, I've had to find a way of reconciling the unsafe nature of the methods on `Numeric` with defining an `Eq`.

My approach has been to require that for two `Numeric` instances to be equal, when one throws the other throws. Lifting the values into `Option` and using `None` to represent cases that throw erases the difference between different types of thrown exceptions. I've decided that this is okay given the alternative would be a dirty `Eq` for `ArithmaticException` and `IllegalArgumentException`.

Happy to change this approach to something else.

Reply:
Actually I think this is unnecessary except in the `fromInt` case. I will rework it...